app/project_type/project_type.py

```diff
@@ -190,17 +194,50 @@ def execute_command_in_project(self, command, extra_environment_vars=None, timeo
         self._logger.debug('Executing command in project: {}', command)
 
         # Redirect output to files instead of using pipes to avoid: https://github.com/box/ClusterRunner/issues/57
-        stdout_file = TemporaryFile()
-        stderr_file = TemporaryFile()
+        output_file = output_file if output_file is not None else TemporaryFile()
         pipe = Popen_with_delayed_expansion(
             command,
             shell=True,
-            stdout=stdout_file,
-            stderr=stderr_file,
+            stdout=output_file,
+            stderr=STDOUT,  # Redirect stderr to stdout, as we do not care to distinguish the two.
             start_new_session=True,  # Starts a new process group (so we can kill it without killing clusterrunner).
             **popen_kwargs
         )
 
+        clusterrunner_error_msgs = self._wait_for_pipe_to_close(pipe, command, timeout)
+        console_output = self._read_file_contents_and_close(output_file)
+        exit_code = pipe.returncode
+
+        if exit_code != 0:
+            max_log_length = 300
+            logged_console_output = console_output
+            if len(console_output) > max_log_length:
+                logged_console_output = '{}... (total output length: {})'.format(console_output[:max_log_length],
+                                                                                 len(console_output))
+
+            # Note we are intentionally not logging at error or warning level here. Interpreting a non-zero return
+            # code as a failure is context-dependent, so we can't make that determination here.
+            self._logger.notice(
+                'Command exited with non-zero exit code.\nCommand: {}\nExit code: {}\nConsole output: {}\n',
+                command, exit_code, logged_console_output)
+        else:
+            self._logger.debug('Command completed with exit code {}.', exit_code)
+
+        exit_code = exit_code if exit_code is not None else -1  # Make sure we always return an int.
+        combined_command_output = '\n'.join([console_output] + clusterrunner_error_msgs)
+        return combined_command_output, exit_code
+
+    def _wait_for_pipe_to_close(self, pipe, command, timeout):
```

Comment:
git/github royally messed up the code diff here. I basically had ripped out a 50-line chunk of code from execute_command_in_project() into a new method, _wait_for_pipe_to_close().